refactor(auto-approve): check first if config is being modified by PR before checking config in repo

sofisl opened the pull request.

chingor13 reviewed (changes requested):
I would decouple the config validation from the logic of approval - it's dangerous to trust anything coming from a PR (you would almost never want to auto-approve a PR modifying the auto-approve configuration anyways).

sofisl commented:
@chingor13, that's a great point! So, this actually won't ever merge something that is changing or creating the auto-approve configuration. Instead, what it does is that it confirms the auto-approve configuration is valid  _before_  merging a given PR (I added this additional check given @tmatsuo's suggestion that someone could ignore a failing check when they initially create an auto-approve configuration, so we could be auto-merging PRs on an invalid configuration).

What this PR does is that it refactors what we check for first. Instead of checking for a configuration in the repository when a PR comes in, it checks to see if an incoming PR is changing the configuration. If it does, then we will just submit a successful/failing check against that configuration, without auto-merging. If there are no changes to the auto-approve configuration, then we check to see if there's a valid auto-approve config in the repository. If that's valid, then we check to see if the PR matches the requirements in the config. If that checks out, then we can merge!

sofisl commented on packages/auto-approve/src/auto-approve.ts:
```diff
@@ -144,82 +132,100 @@ export function handler(app: Probot) {
+        // Then, check to see whether the incoming PR matches the config
+        if (config) {
+          // If config is not valid, this function will submit a failing check, and will not merge the PR
+          const isConfigValid = await validateConfig(
```
Great question! This actually has to do with my response to @chingor13's review. `validateConfig` checks whether the config is valid, creates a check reflecting whether the config is valid, and returns a boolean whether the config is valid. In the first code block (starting at line 146), we're checking whether a pull request is changing the auto-approve.yml config. If it does, then _all_ we need to do is create that check, and that's it -- we explicitly **don't** want to do anything else with that pull request, like approve or merge it. However, if the PR is changing other files other than `auto-approve.yml`, then we want to check if the configuration is valid _and_ that the PR matches the configuration.  This extra check that the configuration is valid comes from @tmatsuo's suggestion to make sure the config is valid before automerging pull requests that match the config, since someone with admin rights or someone who just ignores the failing auto-approve check might merge an invalid config. In that case, we might be automerging pull requests that match a config that we don't allow (e.g., different authors, etc.).

From @chingor13 and your comments, it sounds like this flow might be unclear. Are there any comments/code improvements I could add to make this flow a bit clearer?

SurferJeffAtGoogle replied:
Sounds good.

Please add a a one or two-sentence comment to line 157 explaining why the return value is not examined.

sofisl commented:
@SurferJeffAtGoogle and @chingor13, thanks for the reviews! I've added in comments, renamed the confusing function, and added a test for this specific functionality.
